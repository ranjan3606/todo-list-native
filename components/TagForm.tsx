import React, { useState } from 'react';
import {
  View,
  Text,
  StyleSheet,
  TouchableOpacity,
  TextInput,
  Modal,
  Alert
} from 'react-native';
import { FontAwesome } from '@expo/vector-icons';
import { Colors } from '@/constants/Colors';
import { getTags, saveTags } from '@/services/storage';
import { todoEvents, TODO_EVENTS } from '@/services/eventEmitter';
import { useTranslation } from '@/i18n';

interface TagModalProps {
  visible: boolean;
  colorScheme: 'light' | 'dark';
  mode: 'add' | 'edit';
  tagName: string;
  tagKeywords: string;
  onChangeTagName: (text: string) => void;
  onChangeKeywords: (text: string) => void;
  onSave: () => void;
  onCancel: () => void;
}

export const TagForm = ({
  visible,
  colorScheme,
  mode,
  tagName,
  tagKeywords,
  onChangeTagName,
  onChangeKeywords,
  onSave,
  onCancel
}: TagModalProps) => {
  const { t } = useTranslation();
  const [isSaving, setIsSaving] = useState(false);
  
  const handleSave = async () => {
    if (isSaving) return;
    
    if (!tagName.trim()) {
      Alert.alert(t('common.error'), t('tags.nameRequired'));
      return;
    }
    
    setIsSaving(true);
    
    try {
      // Convert comma-separated string to array and trim each keyword
      const keywords = tagKeywords.split(',')
        .map(k => k.trim())
        .filter(k => k.length > 0);
      
      // Check if any keyword doesn't start with #
      const formattedKeywords = keywords.map(k => 
        k.startsWith('#') ? k : `#${k}`
      );
      
      // Get current tags
      const currentTags = await getTags();
      let updatedTags = { ...currentTags };
      
      // Update tags based on mode
      if (mode === 'add') {
        // Check if tag already exists
        if (updatedTags[tagName]) {
          Alert.alert(t('common.error'), t('tags.alreadyExists', { name: tagName }));
          setIsSaving(false);
          return;
        }
        
        updatedTags[tagName] = formattedKeywords;
        
        // Emit tag added event
        todoEvents.emit(TODO_EVENTS.TAG_ADDED, { name: tagName, keywords: formattedKeywords });
      } else {
        // Edit existing tag
        updatedTags[tagName] = formattedKeywords;
        
        // Emit tag updated event
        todoEvents.emit(TODO_EVENTS.TAG_UPDATED, { name: tagName, keywords: formattedKeywords });
      }
      
      // Save updated tags
      await saveTags(updatedTags);
      
      // Call the original onSave callback
      onSave();
    } catch (error) {
      console.error('Error saving tag:', error);
      Alert.alert(t('common.error'), t('tags.saveError'));
    } finally {
      setIsSaving(false);
    }
  };
  
  return (
    <Modal
      visible={visible}
      transparent={true}
      animationType="slide"
      onRequestClose={onCancel}
    >
      <View style={[styles.modalOverlay, { backgroundColor: Colors[colorScheme].modalBackground }]}>
        <View style={[styles.modalContent, { backgroundColor: Colors[colorScheme].background }]}>
          <Text style={[styles.modalTitle, { color: Colors[colorScheme].text }]}>
            {mode === 'add' ? t('tags.addTag') : t('tags.editTag')}
          </Text>
          
          <TextInput
            style={[styles.input, { color: Colors[colorScheme].text, borderColor: Colors[colorScheme].border }]}
            placeholder={t('tags.tagName')}
            placeholderTextColor={Colors[colorScheme].lightText}
            value={tagName}
            onChangeText={onChangeTagName}
            editable={mode === 'add'} // Only editable in add mode
          />
          
          <TextInput
            style={[styles.input, { color: Colors[colorScheme].text, borderColor: Colors[colorScheme].border }]}
            placeholder={t('tags.keywords')}
            placeholderTextColor={Colors[colorScheme].lightText}
            value={tagKeywords}
            onChangeText={onChangeKeywords}
            multiline={true}
            numberOfLines={3}
          />
          
          <Text style={[styles.helpText, { color: Colors[colorScheme].lightText }]}>
            {t('tags.keywordsHelp')}
          </Text>
          
          <View style={styles.modalActions}>
            <TouchableOpacity 
              style={[styles.modalButton, styles.cancelButton]} 
              onPress={onCancel}
            >
              <Text style={styles.buttonText}>{t('common.cancel')}</Text>
            </TouchableOpacity>
            
            <TouchableOpacity 
              style={[styles.modalButton, styles.saveButton, isSaving && styles.disabledButton]} 
              onPress={handleSave}
              disabled={isSaving}
            >
              <Text style={styles.buttonText}>
                {isSaving ? t('common.saving') : (mode === 'add' ? t('common.save') : t('common.update'))}
              </Text>
            </TouchableOpacity>
          </View>
        </View>
      </View>
    </Modal>
  );
};

const styles = StyleSheet.create({
  modalOverlay: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  modalContent: {
    width: '90%',
    borderRadius: 12,
    padding: 20,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,
    elevation: 5,
  },
  modalTitle: {
    fontSize: 20,
    fontWeight: 'bold',
    marginBottom: 20,
    textAlign: 'center',
  },
  input: {
    borderWidth: 1,
    borderRadius: 8,
    padding: 12,
    marginBottom: 16,
  },
  helpText: {
    fontSize: 12,
    fontStyle: 'italic',
    marginTop: 4,
  },
  modalActions: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginTop: 20,
  },
  modalButton: {
    flex: 1,
    padding: 12,
    borderRadius: 8,
    marginHorizontal: 8,
    alignItems: 'center',
  },
  cancelButton: {
    backgroundColor: '#ccc',
  },
  saveButton: {
    backgroundColor: Colors.primary,
  },
  disabledButton: {
    opacity: 0.6,
  },
  buttonText: {
    color: 'white',
    fontWeight: '600',
  },
});
