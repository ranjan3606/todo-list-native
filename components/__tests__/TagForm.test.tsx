import React from 'react';
import { render, fireEvent, waitFor, act } from '@testing-library/react-native';
import { TagForm } from '../TagForm';

jest.mock('@react-native-async-storage/async-storage', () => ({
  getItem: jest.fn(() => Promise.resolve(null)),
  setItem: jest.fn(() => Promise.resolve(null)),
  removeItem: jest.fn(() => Promise.resolve(null)),
  clear: jest.fn(() => Promise.resolve(null)),
  getAllKeys: jest.fn(() => Promise.resolve([])),
  multiGet: jest.fn(() => Promise.resolve([])),
  multiSet: jest.fn(() => Promise.resolve(null)),
  multiRemove: jest.fn(() => Promise.resolve(null)),
}));

// Mock storage services
jest.mock('@/services/storage', () => ({
  getTags: jest.fn(() => Promise.resolve({})),
  saveTags: jest.fn(() => Promise.resolve())
}));

// Mock event emitter
jest.mock('@/services/eventEmitter', () => ({
  todoEvents: {
    emit: jest.fn()
  },
  TODO_EVENTS: {
    TAG_ADDED: 'tag_added',
    TAG_UPDATED: 'tag_updated'
  }
}));

jest.mock('@/i18n', () => ({
  useTranslation: () => ({
    t: (key: string, params?: any) => params ? `${key}:${JSON.stringify(params)}` : key
  })
}));

jest.mock('@/constants/Colors', () => ({
  Colors: {
    light: { background: '#fff', text: '#000', border: '#ccc', lightText: '#888', modalBackground: '#eee' },
    dark: { background: '#000', text: '#fff', border: '#888', lightText: '#ccc', modalBackground: '#222' },
    primary: '#007bff',
  }
}));

describe('TagForm', () => {
  const baseProps = {
    visible: true,
    colorScheme: 'light' as const,
    mode: 'add' as const,
    tagName: '',
    tagKeywords: '',
    onChangeTagName: jest.fn(),
    onChangeKeywords: jest.fn(),
    onSave: jest.fn(),
    onCancel: jest.fn(),
  };

  it('renders correctly (snapshot)', () => {
    const { toJSON } = render(<TagForm {...baseProps} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('calls onCancel when cancel button is pressed', async () => {
    const onCancel = jest.fn();
    const { getByText } = render(<TagForm {...baseProps} onCancel={onCancel} />);
    
    await act(async () => {
      fireEvent.press(getByText('cancel'));
    });
    
    expect(onCancel).toHaveBeenCalled();
  });

  it('calls onSave when save button is pressed', async () => {
    const onSave = jest.fn();
    const { getByText } = render(<TagForm {...baseProps} tagName="tag1" tagKeywords="kw1,kw2" onSave={onSave} />);
    
    await act(async () => {
      fireEvent.press(getByText('save'));
    });
    
    // The act() already waits for state updates, so we can simply check
    expect(onSave).toHaveBeenCalled();
  });

  it('renders in edit mode', () => {
    const { getByDisplayValue } = render(<TagForm {...baseProps} mode="edit" tagName="tag1" tagKeywords="kw1,kw2" />);
    expect(getByDisplayValue('tag1')).toBeTruthy();
    expect(getByDisplayValue('kw1,kw2')).toBeTruthy();
  });
});
