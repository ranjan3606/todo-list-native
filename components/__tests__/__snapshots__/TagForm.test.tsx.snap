// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagForm renders correctly (snapshot) 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[MockFunction]}
  transparent={true}
  visible={true}
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        },
        {
          "backgroundColor": "#eee",
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "borderRadius": 12,
            "elevation": 5,
            "padding": 20,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.25,
            "shadowRadius": 3.84,
            "width": "95%",
          },
          {
            "backgroundColor": "#fff",
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "fontSize": 20,
              "fontWeight": "bold",
              "marginBottom": 20,
              "textAlign": "center",
            },
            {
              "color": "#000",
            },
          ]
        }
      >
        tags.addTag
      </Text>
      <TextInput
        editable={true}
        onChangeText={[MockFunction]}
        placeholder="tags.tagName"
        placeholderTextColor="#888"
        style={
          [
            {
              "borderRadius": 8,
              "borderWidth": 1,
              "marginBottom": 16,
              "padding": 12,
            },
            {
              "borderColor": "#ccc",
              "color": "#000",
            },
          ]
        }
        value=""
      />
      <TextInput
        multiline={true}
        numberOfLines={3}
        onChangeText={[MockFunction]}
        placeholder="tags.keywords"
        placeholderTextColor="#888"
        style={
          [
            {
              "borderRadius": 8,
              "borderWidth": 1,
              "marginBottom": 16,
              "padding": 12,
            },
            {
              "borderColor": "#ccc",
              "color": "#000",
            },
          ]
        }
        value=""
      />
      <Text
        style={
          [
            {
              "fontSize": 12,
              "fontStyle": "italic",
              "marginTop": 4,
            },
            {
              "color": "#888",
            },
          ]
        }
      >
        tags.keywordsHelp
      </Text>
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 20,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#ccc",
              "borderRadius": 8,
              "flex": 1,
              "marginHorizontal": 8,
              "opacity": 1,
              "padding": 12,
            }
          }
        >
          <Text
            style={
              {
                "color": "white",
                "fontWeight": "600",
              }
            }
          >
            cancel
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#007bff",
              "borderRadius": 8,
              "flex": 1,
              "marginHorizontal": 8,
              "opacity": 1,
              "padding": 12,
            }
          }
        >
          <Text
            style={
              {
                "color": "white",
                "fontWeight": "600",
              }
            }
          >
            save
          </Text>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
